### PR TITLE
Add: $removeCooldown, $wordCount and $editEmbed

### PR DESCRIPTION
**Pull request by SaiaAmako5**
## Type
- [ ] Bug Fix
- [x] Functions: $removeCooldown, $editEmbed and $wordCount
- [ ] Callbacks: \<Callback Name>
- [ ] Handlers: \<Handler Type>
- [ ] Others: \______

Dependencies (Third Party Modules) needed: None. Not needed. 

Want a credit? Discord tag or other social media link: saiaamako | [saiaamako5](https://github.com/SaiaAmako5) 

Referenced Issue: #NaN (Answer for an issue, if any)

## Description
Adding three functions: $removeCooldown, $editEmbed and $wordCount that i made.

### CODE REVIEW

File: src/functions/guild/editEmbed.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
+module.exports = async d => {
```
Comment (Faf4a) [inline src/functions/guild/editEmbed.js:1]: Use prettier and follow the project's styling.

File: src/functions/guild/editEmbed.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+const fi = option.charAt(6) - 1;
+option = option.slice(0, 6) + fi + option.slice(7); 
+option = option.slice(0, 5) + "s" + option.slice(5); 
+} else if(option.startsWith("field")) {
+const fi = option.charAt(5) - 1;
+option = option.slice(0, 5) + fi + option.slice(6); 
+option = option.slice(0, 5) + "[" + option.slice(5);
+option = option.slice(0, 7) + "]" + option.slice(7);
+option = option.slice(0, 5) + "s" + option.slice(5); 
+} else if(option == "timestamp") {
```
Comment (Faf4a) [inline src/functions/guild/editEmbed.js:25]: don't do this, this is actually unreadable, check parsers src for reference

File: src/functions/util/wordCount.js [hunk 3]
```diff
@@ -0,0 +1,13 @@
+
```
Comment (Faf4a) [inline src/functions/util/wordCount.js:1]: follow styling here as well

File: src/functions/guild/editEmbed.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  const data = d.util.aoiFunc(d);
+  var [channelID = d.channel?.id, messageID = d.message?.id, index = 1, option = 'description', content] = data.inside.splits;
+ index = index - 1;
+const channel = await d.util.getChannel(d, channelID);
+if (!channel) return d.aoiError.fnError(d, 'channel', {}, {inside: data.inside});
+const message = await d.util.getMessage(channel, messageID);
+if (!message) return d.aoiError.fnError(d, 'message', {}, {inside: data.inside});
+function isNumber(value) {
+  return typeof value === "number" && !Number.isNaN(value);
+}
```
Comment (Faf4a) [inline src/functions/guild/editEmbed.js:11]: just use isNaN(Number), aoijs uses the concept of strings therefore numbers dont exist unless converted into using Number() or similar iirc

**Review by Faf4a — CHANGES_REQUESTED** [review]
Thanks for the pull request and the useful additions, however think about applying the suggestions!

File: src/functions/guild/editEmbed.js (hunk 2 above)
Comment (SaiaAmako5) [reply src/functions/guild/editEmbed.js]: Ah i am sorry, I don't know why I did this, when I could have just used .substring(). Well I committed a new update for it. Please look into it.

File: src/functions/guild/editEmbed.js (hunk 4 above)
Comment (SaiaAmako5) [reply src/functions/guild/editEmbed.js]: Okay

File: src/functions/guild/editEmbed.js (hunk 1 above)
Comment (SaiaAmako5) [reply src/functions/guild/editEmbed.js]: Used prettier. Please check it.

File: src/functions/editEmbed.js [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+      .replaceAll("footertext", "footer.text")
+      .replaceAll("footericon", "footer.icon_url");
+    if (option == "image" || option == "thumbnail") {
+      option = option
+        .replaceAll("image", "image.url")
+        .replaceAll("thumbnail", "thumbnail.url");
+    }
+  }
+  try {
+    if (!eval(`embed.data.${option}`))
```
Comment (Faf4a) [inline src/functions/editEmbed.js:46]: Please prevent using eval() directly, unless absolutely needed.

File: src/functions/editEmbed.js [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+      return d.aoiError.fnError(d, "custom", {}, "Invalid Option Provided");
+    eval(`embed.data.${option} = content.addBrackets()`);
+  } catch {
+    d.aoiError.fnError(d, "custom", {}, "Option/Index Not Found");
+    return;
```
Comment (Faf4a) [inline src/functions/editEmbed.js:51]: Merge those into the same line

File: src/functions/editEmbed.js [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+  if (option.startsWith("field")) {
+    const fi = option.substring(option.indexOf("field") + 1).match(/\d+/) - 1;
+    option = "fields[" + fi + "]." + option.substring(option.indexOf(".") + 1);
+  } else if (option == "timestamp") {
+    try {
+      const ts = new Date(Number(content));
+      content = ts.toISOString();
+    } catch {
+      d.aoiError.fnError(d, "custom", {}, "Invalid MS Provided");
+      return;
```
Comment (Faf4a) [inline src/functions/editEmbed.js:31]: Merge those in the same line

File: src/functions/editEmbed.js [hunk 8]
```diff
@@ -0,0 +1,59 @@
+module.exports = async (d) => {
+  const data = d.util.aoiFunc(d);
+  var [
+    channelID = d.channel?.id,
+    messageID = d.message?.id,
+    index = 1,
+    option = "description",
```
Comment (Faf4a) [inline src/functions/editEmbed.js:7]: Remove default option, probably not the best to have this especially when some embeds dont have a description

File: src/functions/editEmbed.js [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+  ] = data.inside.splits;
+  index = index - 1;
+  const channel = await d.util.getChannel(d, channelID);
+  if (!channel)
+    return d.aoiError.fnError(d, "channel", {}, { inside: data.inside });
+  const message = await d.util.getMessage(channel, messageID);
+  if (!message)
+    return d.aoiError.fnError(d, "message", {}, { inside: data.inside });
+  if (isNaN(index))
+    return d.aoiError.fnError(d, "custom", {}, "Invalid Index");
```
Comment (Faf4a) [inline src/functions/editEmbed.js:18]: this doesnt actually check for negative numbers, add a check for that

File: src/functions/editEmbed.js [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+  if (!content)
+    return d.aoiError.fnError(d, "custom", {}, "No Content Provided");
+  var embed = new d.embed(message.embeds[index]);
```
Comment (Faf4a) [inline src/functions/editEmbed.js:21]: dont use var, use let

File: src/functions/editEmbed.js [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+    if (option == "image" || option == "thumbnail") {
+      option = option
+        .replaceAll("image", "image.url")
+        .replaceAll("thumbnail", "thumbnail.url");
+    }
+  }
+  try {
+    if (!eval(`embed.data.${option}`))
+      return d.aoiError.fnError(d, "custom", {}, "Invalid Option Provided");
+    eval(`embed.data.${option} = content.addBrackets()`);
```
Comment (Faf4a) [inline src/functions/editEmbed.js:48]: try to not use eval

File: src/functions/editEmbed.js
Comment (Faf4a) [inline src/functions/editEmbed.js:1]: Use the prettier config inside the root directory, prettier should detect it when using `ALT + SHIFT + F` or `prettier --w`!

**Review by Faf4a — CHANGES_REQUESTED** [review]
*Few* changes before this can get merged! Let me know if you're still interested on working on these functions, sorry for the long wait!

**Review by Faf4a — APPROVED** [review]
LGTM

File: src/functions/removeCooldown.js [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+    const data = d.util.aoiFunc(d);
+    const [command, type, ...id] = data.inside.splits;
+    const types = {
+        globalUser: id[0] || d.message.author.id,
+        user: `${id[0] || d.message.author.id}_${id[1] || d.message.guild.id || "dm"}`,
+        guild: id[0] || d.message.guild.id,
+        channel: id[0] || d.message.channel.id
+    };
+    if (!types[type]) return d.aoiError.fnError(d, "custom", { inside: data.inside }, "Invalid Type In");
+    if (await d.client.db.get(type == "guild" ? d.client.db.tables[0] : "__aoijs_vars__", "cooldown", command + "_" + types[type])) {
```
Comment (Faf4a) [inline src/functions/removeCooldown.js:13]: Why?

Cooldowns are always stored into the internal `__aoijs_vars__`.

**Review by Faf4a — CHANGES_REQUESTED** [review]

File: src/functions/removeCooldown.js (hunk 12 above)
Comment (SaiaAmako5) [reply src/functions/removeCooldown.js]: You can check the `guildCooldown.js`, it seems to store the CD in the first db table instead of `__aoijs_vars__`.

Comment (Faf4a) [reply src/functions/removeCooldown.js:13]: That's indeed unintended behaviour, which will be fixed in the next version. Therefore please change this to the `__aoijs_vars__` table.

**Review by Leref — COMMENTED** [review]
I mean it's good beside @Faf4a stated of variable format.

File: src/functions/removeCooldown.js (hunk 12 above)
Comment (SaiaAmako5) [reply src/functions/removeCooldown.js]: I see. I have committed a change. Please check it if it's okay.

**Comment by Leref** [comment]
Will close since no updates. Re-open if wish to maintain this PR.

**Pull request closed without merging**